### PR TITLE
Extended Moderation of comments

### PR DESCRIPTION
**Pull request by ksuess**
Existing review workflow extended by two new workflow states "rejected" and "spam".
Moderation view displays comments with filtering by workflow state.
image

### CODE REVIEW

**Comment by ksuess** [comment]
@jenkins-plone-org please run jobs

**Comment by ksuess** [comment]
collective/plone.app.locales#285 adds the necessary translations.

So tests below run with
#166
collective/plone.app.locales#285

plus
plone/plone.restapi#843
plone/Products.CMFPlacefulWorkflow#30
plone/Products.CMFPlone#3002
for failing tests on additional workflow states.

```
https://github.com/plone/plone.app.discussion/pull/166
https://github.com/collective/plone.app.locales/pull/285
https://github.com/plone/plone.restapi/pull/843
https://github.com/plone/Products.CMFPlacefulWorkflow/pull/30
https://github.com/plone/Products.CMFPlone/pull/3002
```

**Comment by mauritsvanrees** [comment]
Just very quickly looking at this. Don't we simply want to delete comments instead of marking them as spam or rejected? When you only mark them, they stay in the database forever.

Are these new states options that you can enable or disable in the control panel?

**Comment by ksuess** [comment]
That's the point: Not published comments stay and are still accessible. And the moderator can push them out of the default view of pending comments by marking them as rejected or spam. See filter on top of screenshot above.

Of course the moderator can still delete single comments like before and he also can batch delete multiple comments.

**Comment by ksuess** [comment]
There is still one test for Plone 5.1 failing:
```
    browser.getControl(name='form.widgets.IRichTextBehavior.text').value = "Lorem ipsum"
    LookupError: name 'form.widgets.IRichTextBehavior.text'
```
This is because plone.app.contenttypes is checked out with branch 1.4.x, but the renaming of the interface from IRichText to IRichTextBehavior happend in version 2.0.0
What about changing to master branch?

**Comment by mauritsvanrees** [comment]
Current Plone 5.1 is not using the master branch of `plone.app.discussion`. It is using a branch that sadly is called `3.x`, but that *should* have been called `3.0.x`. The last release on that branch was 3.0.8, which is what is in the `versions.cfg`. That may explain several test failures.

So on Plone 5.1 this would need a different PR.
Or actually preferably no PR at all: it seems best to do this big new feature only on Plone 5.2.

**Comment by ale-rt** [comment]
It also seems to me something 5.2 specific (I did not test it anyway, will do now)

File: plone/app/discussion/browser/moderation.py [hunk 1]
```diff
@@ -13,69 +15,119 @@
...
+_('comment_rejected', default='rejected')
+_('comment_spam', default='marked as spam')
+# Transitions
+_('Recall')
+_('Approve')
+_('Reject')
+_('Spam')
+
+
+class TranslationHelper(BrowserView):
```
Comment (ale-rt) [inline plone/app/discussion/browser/moderation.py]: I am not really sure this is needed. I would avoid introducing this new view here.
AFAICS this is used only in `CommentTransition`. 
Would it be possible to move those methods there or you need this as an entry point to provide custom translations?

File: plone/app/discussion/browser/moderation.py [hunk 2]
```diff
@@ -13,69 +15,119 @@
...
+
+    def translate(self, text=""):
```
Comment (ale-rt) [inline plone/app/discussion/browser/moderation.py]: I might be wrong but the method `translate` does not seem to be used.

File: plone/app/discussion/browser/moderation.py [hunk 3]
```diff
@@ -13,69 +15,119 @@
...
-        catalog = getToolByName(context, 'portal_catalog')
-        self.comments = catalog(object_provides=IComment.__identifier__,
-                                review_state='pending',
-                                sort_on='created',
-                                sort_order='reverse')
+        self.request.set('review_state',
+                         self.request.get('review_state', 'pending'))
         return self.template()
 
+    def comments(self):
```
Comment (ale-rt) [inline plone/app/discussion/browser/moderation.py]: Making this a `@property` will not break the previous behavior.

File: plone/app/discussion/browser/moderation.py [hunk 4]
```diff
@@ -13,69 +15,119 @@
...
+
     template = ViewPageTemplateFile('moderation.pt')
     try:
         template.id = '@@moderate-comments'
     except AttributeError:
         # id is not writeable in Zope 2.12
         pass
 
+    def __init__(self, context, request):
+        super(View, self).__init__(context, request)
```
Comment (ale-rt) [inline plone/app/discussion/browser/moderation.py]: I would avoid customizing the `__init__` of the BrowserView, especially with mutable objects (the `transitions` list).
It makes hard to understand what happens under the hood and might lead to unexpected results.

File: plone/app/discussion/browser/moderation.py [hunk 5]
```diff
@@ -256,58 +312,53 @@ class BulkActionsView(BrowserView):
 
     """
 
+    def __init__(self, context, request):
+        super(BulkActionsView, self).__init__(context, request)
+        self.workflowTool = getToolByName(context, 'portal_workflow')
```
Comment (ale-rt) [inline plone/app/discussion/browser/moderation.py]: I would not redefine the init of the view.

**Comment by ale-rt** [comment]
@ksuess @mauritsvanrees I think that with this https://github.com/plone/Products.CMFPlone/pull/3036 it will be easier merging this one.

**Comment by mauritsvanrees** [comment]
There was a bit of a discussion at the Alpine City Sprint on whether we should include new features like this in a bugfix release of 5.2. For stability it would be better if 5.2 only gets bug fixes, not new features. That is certainly true for Plone 5.1 (the failing Jenkins 5.1 job can be ignored: 5.1 uses another branch anyway, and should not have been tested with this).

But at the same time we are busy in 5.2 to update Zope from 4.1 to 4.2/4.3.  That is more likely to have unintended side effects than this new feature in `plone.app.discussion`.

So while I like stability, I think for 5.2 we can allow this PR.

Anyone else want to comment, especially the @plone/framework-team ?

Looks like this needs to be merged with https://github.com/plone/plone.restapi/pull/843. Also with https://github.com/collective/plone.app.locales/pull/285, but I don't think this will cause test failures when it is merged a bit later.

**Comment by jensens** [comment]
From my perspective I am +1 for merging, unless we change our versioning policy - we had similar complex changes already in x.y.Z releases and this is a useful one. Plone major release versions do not follow Semver but are marketing versions. This was once decided at a PLOG (and I am not happy with this decision). So, with the rule in place we can add little features on Z and not feel bad.

**Comment by mauritsvanrees** [comment]
Okay, we can conclude it is approved for inclusion in Plone 5.2.
Katja, when I have it correct, this means that we should do things in this order:

1. Merge https://github.com/plone/plone.restapi/pull/843 (@plone/restapi-team should do that). That is just a minor test fix that is fine even without the current `plone.app.discussion` PR. It must be merged first though.
2. Merge the current PR.
3. Merge the translations: https://github.com/collective/plone.app.locales/pull/285

Then it is done, right?

**Comment by ksuess** [comment]
@ale-rt thank you for your effort on all versions "Improve tests for the workflow tool method listWFStatesByTitle". A hero point from me...
@mauritsvanrees thank you for your special effort on caring about this multiple package thing. 
Your comment https://github.com/plone/plone.app.discussion/pull/166#issuecomment-598662157
 does reduce the topic to two package PRs, hoohray.
That's it. That would push comments moderation forward.
Thank you all, @ale-rt , @mauritsvanrees  and you, not mentioned but involved helping hands. 
Keep healthy!

**Comment by ale-rt** [comment]
Thanks to **you** for making Plone better!

**Comment by mauritsvanrees** [comment]
I have merged the PR. Congrats and again thanks!
The `plone.app.locales` PR is open, but I have asked Vincent Fretin to have a look at it first, otherwise we will lose perfectly fine translation on Plone 5.1.

**Comment by ksuess** [comment]
https://github.com/plone/plone.app.discussion/pull/166#issuecomment-598662157
Translation seems to be OK for merging.

**Comment by ksuess** [comment]
And, yes, I'll come back to your comments @ale-rt

**Pull request merged**